File: built-in plugins dsp/mb-stereo-imager-ms.cpp
/**
 * MB MS Stereo Imager
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Mid-side stereo image processor
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STEREO_IMAGER_MS_H
#define MB_STEREO_IMAGER_MS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStereoImagerMs : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-stereo-imager-ms";
    static constexpr const char* PLUGIN_NAME    = "MB MS Stereo Imager";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float midGain = 0f;  // range [-24, 24]
    float sideGain = 0f;  // range [-24, 24]
    float balance = 0f;  // range [-1, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbStereoImagerMs() = default;
    ~MbStereoImagerMs() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.midGain = std::clamp(params.midGain, -24f, 24f);
        params.sideGain = std::clamp(params.sideGain, -24f, 24f);
        params.balance = std::clamp(params.balance, -1f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB MS Stereo Imager
        return input;
    }
};

#endif // MB_STEREO_IMAGER_MS_H

File: built-in plugins dsp/mb-stereo-imager-ms.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbStereoImagerMsPlugin: PluginDefinition = { id: 'mb-stereo-imager-ms', slug: 'mb-stereo-imager-ms', name: 'MB MS Stereo Imager', category: 'effect', type: 'stereo' as any, version: '1.0.0', description: 'Mid-side stereo image processor', author: 'Max Booster', grade: 'A', parameters: [{ id: 'midGain', name: 'Mid Gain', type: 'float', defaultValue: 0, minValue: -24, maxValue: 24, automatable: true }, { id: 'sideGain', name: 'Side Gain', type: 'float', defaultValue: 0, minValue: -24, maxValue: 24, automatable: true }, { id: 'balance', name: 'Balance', type: 'float', defaultValue: 0, minValue: -1, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { midGain: 0, sideGain: 0, balance: 0, mix: 1 } };

export default MbStereoImagerMsPlugin;
